### DISCUSSION TITLE
Is there a tutorial for upgrading from version 5 to 6?

### DISCUSSION THREAD

**Discussion by SyaoranChang** [LanKuDot/game_modules#43] [Q&A] [answered]
Hi,

I want to upgrade from version 5 to 6.
ListPositionCtrl.cs file seems to have disappeared.
ListPositionCtrl.GetCenteredBox() is gone, What to use instead?
Is the GetFocusingBox() to use instead?

ListBox.cs same  funtion was disappeared too.
public virtual void UpdatePosition(float delta) is gone.

Currently do not know how to deal with the version of the code written in version 5, upgrade to version 6

**Comment by LanKuDot** [comment, marked as answer]
Yeah, because of the introduction of the top and bottom focusing position.
All `___Centered___` related functions or events have been renamed to `___Focusing___`:

- `GetCenteredBox()` -> `GetFocusingBox()`
- `GetCenteredContentID()` -> `GetFocusingContentID()`
- event `OnCenteredContentChanged` -> `OnFocusingBoxChanged`
- event `OnBoxClick` -> `OnBoxSelected`
- option `CenteredContentID` -> `InitFocusingContentID`
- option `CenterSelectedBox` -> `FocusSelectedBox`

And for the `UpdatePositon()`, it is not a virtual function in the version 5. Maybe there has the modified code.
The code structure has been modified in the version 6.
Therefore, we could start from what kind of effect you want to achieve.
And then find the solution for the version 6.

**Comment by SyaoranChang** [comment]
Hi,

I did not write the initial code changes, but seems want to do like to change the background from gray to color when handling an object from unfocused to focusing, and to change the background from colored to gray when focusing to unfocused in UpdatePositon().

**Reply by LanKuDot** [reply]
Got it, I think you could use `OnFocusingBoxChanged` event to get the previous and the current focusing box.
The type of the argument is `ListBox`, and it could be converted back to the box type you defined.

**Comment by SyaoranChang** [comment]
Hi,

1.How to set ScrollingList.ListSetting.InitFocusingContentID by script?
Seems InitFocusingContentID  is read only.

I try ScrollingList.SelectContentID() or ScrollingList.Refresh() will get null error If called too early when ScrollingList is not Initialized.

2.~~How to avoid null issue when call ScrollingList.SelectContentID() or ScrollingList.Refresh()?
Seems CheckIsInitialized() is private, can not use to avoid.~~
I use call StartCoroutine(LateStart()) in Start() to solve it.
```
IEnumerator LateStart()
{
        yield return null;
        foreach (ListBox box in m_ScrollingList.ListBoxes)
        {
            m_ScrollingList.Refresh(box.ContentID); 
        }
}
```

About Refresh(int focusingContentID = -1)
Seems version 5 will check centeredContentID < 0 then give 0.
But version 6 will throw new IndexOutOfRangeException.
version 5:
```
if (centeredContentID < 0)
                centeredContentID =
                    centeredBox.contentID == int.MinValue
                        ? 0
                        : Mathf.Min(centeredBox.contentID, numOfContents - 1);
            else if (centeredContentID >= numOfContents)
                throw new IndexOutOfRangeException(
                    $"{nameof(centeredContentID)} is larger than the number of contents");
```
version 6
```
            if (!_listContentProvider.IsIDValid(contentID))
                throw new IndexOutOfRangeException(
                    $"'{nameof(contentID)}' is invalid");
```
3.How to simply update all the ListBox, or can only call like this
```
foreach (ListBox box in m_ScrollingList.ListBoxes)
        {
            m_ScrollingList.Refresh(box.ContentID); 
        }
```

**Reply by LanKuDot** [reply]
For 1 and 2,

The list is automatically initialized in `Start()` of the `CircularScrollingList.cs`.
You could close the "Initialize On Start" option and invoke the `Initialize()` manually to control when to initialize it.
Therefore, if you invoke `Refresh()` or `SelectionContentID()` before the list is initialized, it will throw the exception, because the list haven't retrieve the content from the bank yet.

For 3,

To refresh content of the box, just invoke `CircularScrollingList.Refresh()` (without specifying the content id) in one call.
All the boxes will reacquire the content from the bank according to its current content ID,

**Reply by SyaoranChang** [reply]
Hi,

I try to close the "Initialize On Start" option and invoke the Initialize() on my UI OpenDone().
But after UI Closed and reopen again, I don't have the information for me to check if I need to call Initialize();

**Reply by LanKuDot** [reply]
No need to initialize again. The `CircularScrollingList.cs` component is ready once the `Initialize()` is called.

**Reply by LanKuDot** [reply]
By the way, if you want to check if the list is initialized. Add the following code to the `CircularScrollingList.cs`:

```csharp
public bool IsInitialized => _isInitialized;
```

**Reply by SyaoranChang** [reply]
Thank you for your answer, so that I can upgrade smoothly.
